x-pack/plugin/ingest/src/main/java/org/elasticsearch/xpack/ingest/FingerprintProcessor.java

```diff
@@ -0,0 +1,295 @@
+import java.util.Comparator;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.ingest.ConfigurationUtils.newConfigurationException;
+import static org.elasticsearch.ingest.ConfigurationUtils.readBooleanProperty;
+
+public final class FingerprintProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "fingerprint";
+
+    static final byte[] DELIMITER = new byte[] { 0 };
+    static final byte[] TRUE_BYTES = new byte[] { 1 };
+    static final byte[] FALSE_BYTES = new byte[] { 2 };
+
+    private final List<String> fields;
+    private final String targetField;
+    private final ThreadLocal<Hasher> threadLocalHasher;
+    private final byte[] salt;
+    private final boolean ignoreMissing;
+
+    FingerprintProcessor(
+        String tag,
+        String description,
+        List<String> fields,
+        String targetField,
+        byte[] salt,
+        ThreadLocal<Hasher> threadLocalHasher,
+        boolean ignoreMissing
+    ) {
+        super(tag, description);
+        this.fields = new ArrayList<>(fields);
+        this.fields.sort(Comparator.naturalOrder());
+        this.targetField = targetField;
+        this.threadLocalHasher = threadLocalHasher;
+        this.salt = salt;
+        this.ignoreMissing = ignoreMissing;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+        Hasher hasher = threadLocalHasher.get();
+        hasher.reset();
+        hasher.update(salt);
+
+        var values = new Stack<>();
+        for (int k = fields.size() - 1; k >= 0; k--) {
+            String field = fields.get(k);
+            Object value = ingestDocument.getFieldValue(field, Object.class, true);
+            if (value == null) {
+                if (ignoreMissing) {
+                    continue;
+                } else {
+                    throw new IllegalArgumentException("missing field [" + field + "] when calculating fingerprint");
+                }
+            }
+            values.push(value);
+        }
+
+        if (values.size() > 0) {
+            // iteratively traverse document fields
+            while (values.isEmpty() == false) {
+                var value = values.pop();
+                if (value instanceof List) {
+                    var list = (List<?>) value;
+                    for (int k = list.size() - 1; k >= 0; k--) {
+                        values.push(list.get(k));
+                    }
+                } else if (value instanceof Set) {
+                    var set = (Set<?>) value;
+                    // process set entries in consistent order
+                    var setList = set.stream().sorted().collect(Collectors.toList());
+                    for (int k = setList.size() - 1; k >= 0; k--) {
+                        values.push(setList.get(k));
+                    }
```

Comment:
Streams are super great, but they aren't quite as fast compared to `foreach` clauses yet, and since this is performance sensitive code, what do you think about doing:

 ```suggestion
                    var setList = new ArrayList<?>(set);
                    setList.sort(Comparator.naturalOrder());
                    for (Object thing : setList) {
                        values.push(thing);
                    }
```

?

(I might be missing something about why you're using a for loop indexed rather than `foreach`, as I *think* the `foreach` compiles to the same bytecode, but I could totally be wrong)

Reply:
The for loop iterates backward over the sorted list to push the values onto the stack. That means that the values are processed in sorted order which is not strictly necessary since only a stable hash is required, but having the values in sorted order is certainly nicer when debugging.

Reply:
I did replace the usage of streams both here and below.